main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1231,6 +1231,44 @@ object Typer extends Phase[ResolvedAst.Program, TypedAst.Root] {
           resultEff <- unifyEffM(evar, eff1, eff2, loc)
         } yield (resultTyp, resultEff)
 
+      case ResolvedAst.Expression.FixpointFold(constraints, init, f, tvar, evar, loc) =>
+        //
+        // constraints : #{a}    init : b   f : a' -> b -> b
+        // where a' is the tuple reification of relation a
+        // ---------------------------------------------------
+        // fold constraints init f : b
+        //
+        // TODO: this has to be cleaned
+        def tupleForSchema(schemaType: Type): InferMonad[Type] = {
```

Comment:
Yikes!

I have a suggestion:

We need the predicate symbol as part of the fold.

Given the symbol, you can look up its arity `n`.

Now given `n`, you can construct type variables `t1` to `tn`.

Now require the function has type `(acc, (t1, ..., tn)) -> ` and require that the constraint set contains the row type `P(t1, ..., tn`.

Does that work?

Reply:
i.e. fold <Predicate> <FunctExp> <InitExp> <ConstraintSetExp>

Reply:
Sounds good! Indeed having the predicate symbol explicit will clean things up and be more consistent with `project`.